lib/std/heap.zig

```diff
@@ -8,6 +8,129 @@ const c = std.c;
 const Allocator = std.mem.Allocator;
 const windows = std.os.windows;
 
+const page_size_darwin: ?comptime_int = if (builtin.os.tag.isDarwin())
+    switch (builtin.cpu.arch) {
+        .x86, .x86_64 => 4 << 10,
+        .thumb, .thumbeb, .arm, .armeb, .aarch64, .aarch64_be, .aarch64_32 => 16 << 10,
+        else => null,
+    }
+else
+    null;
+
+// -- <https://devblogs.microsoft.com/oldnewthing/20210510-00/?p=105200>
+const page_size_windows: ?comptime_int = if (builtin.os.tag == .windows and builtin.os.version_range.windows.min.isAtLeast(.xp))
```

Comment:
You can drop this check; the Zig standard library only supports Windows 8.1+: https://github.com/ziglang/zig/blob/a69d403cb2c82ce6257bfa1ee7eba52f895c14e7/lib/std/Target.zig#L495-L500

Reply:
https://github.com/ziglang/zig/blob/a69d403cb2c82ce6257bfa1ee7eba52f895c14e7/lib/std/Target.zig#L340-L342

https://github.com/ziglang/zig/blob/a69d403cb2c82ce6257bfa1ee7eba52f895c14e7/lib/std/Target.zig#L348-L350

As far as I can tell, that doc comment means the standard library is guaranteed to support features from Windows 8.1 onward, not that it shouldn't support any features outside that range (though, only Andrew, who introduced it in fba39ff331a84f1a32d076ccbb8b87cd02ea7121, knows). With something as fundamental as page size, maybe it's worth supporting older versions of Windows when trivial, like in this case?

Reply:
It really is our minimum. For example, we recently removed fallback code for Linux kernels older than 4.19 in the standard library because that's the oldest version we claim to support in `Target.zig`.

Reply:
That's a shame, thanks for clearing that up. I don't see how we can claim to be a successor to C when we fail to support a similar number of targets.

Reply:
Well, Microsoft themselves do not support Windows versions that old. In fact, Windows 8.1 even reached EoL over a year ago, so there's a very strong argument to be made that we should move to 10+ for Zig now. We similarly follow Apple's support policy for macOS, and Linux 4.19 is the oldest still-supported LTS kernel (expected to EoL at the end of 2024, after which we presumably move to 5.4).

This isn't to say that you can't use the Zig compiler to target older (insecure!) platforms, it's just that the standard library abstractions specifically won't support them.

Reply:
> it's just that the standard library abstractions specifically won't go out of their way to support them.

Well, that's the thing, right? It's virtually no effort to add two switch prongs and one if statement -- we're certainly not "going out of our way" here, but are actively removing support that was already present. One might argue that supporting these platforms isn't worthwhile, but since they are still used in practice and Zig's main goal is to be more pragmatic than C, I think there's a case to be made for supporting them. Either way, I'm getting off-topic, this isn't the place to discuss stdlib OS support. Thanks for your clarification ^_^

Reply:
I'm just echoing the decision that was already made before I started seriously contributing to Zig, but I think it's reasonable to have the debate; I would suggest opening a separate issue. (If you do, I can also elaborate there on what motivated me to remove code for Linux kernels older than 4.19.)

Reply:
Not sure where the last public statement was, the one I remembered is here: https://github.com/ziglang/zig/issues/7242#issuecomment-736772022

> Even without explicit support, non invasive patches to widen support are welcome. They won't be tested by the CI though and will be maintained on best effort, but without mandatory testing.

As long as someone is willing to put forward the effort, and it doesn't overly complicate code, I believe such support is welcome. (But feel free to correct me if this has changed since that comment.)

Reply:
Four years is a lifetime in Zig development terms. 🙂 I believe the decision to follow Apple support policy [came a couple of years after that comment](https://ziglang.org/download/0.10.0/release-notes.html#macOS), and then recently Andrew okayed dropping old Linux kernel support in #20389. Given these facts, I don't think we should necessarily assume that Andrew's position on this remains the same without confirmation - and especially not for something as dead as Windows XP.

Reply:
> and then recently Andrew okayed dropping old Linux kernel support in https://github.com/ziglang/zig/pull/20389.

I assume you mean [the discussion around `statx`](https://github.com/ziglang/zig/pull/20389#discussion_r1685677972) - as an outsider, to me that particular exchange reads like trying to find a suitable solution to an open design question. Your suggestion of dropping support and having a cleaner single code path that doesn't support the older versions aligns with the goals of `std`, so it was accepted in this case. That's just how this particular conversation resolved; I wouldn't necessarily use that as a reason to remove support for older versions in other scenarios. (Also, I'm not trying to pin the decision on you, sorry if it sounds that way.)

As long as the respective commits/PRs gain something from it (like simplicity), removing legacy support may be worthwhile to streamline the development of Zig/`std`.
We are also always free to restore support for older versions later on. This best-effort legacy support also doesn't need to be universal - we can have it for some common file system operations but not include others. It's really just a low-priority extra that a rather small group of users might appreciate.

AFAIU ["dropping support" as stated in the release notes](https://ziglang.org/download/0.10.0/release-notes.html#macOS) means the core contributors aren't interested in investing time in it; I don't think it means easy-to-merge fixes won't be integrated. (Though of course I may be wrong.)
Also wanted to note that the commit https://github.com/ziglang/zig/commit/0d4d8dfc1583138a59147f1aab5eec803a23adb1 referenced in https://github.com/ziglang/zig/issues/20423 adjusts the "default version minimum", i.e. the lower bound of the os version range when no such range is specified in the target triple via the CLI. It doesn't raise the minimum version that can be specified in the target triple. Similarly I don't understand https://github.com/ziglang/zig/issues/20423 to mean "we don't want to support older versions", only "we don't care about / won't promise supporting older versions".

Sorry for the giant walls of text, I'm just trying to be thorough. And thank you for your references - until Andrew swoops in to clear it up, we can all just constructively posit our best guesses.

Reply:
> Your suggestion of dropping support and having a cleaner single code path that doesn't support the older versions aligns with the goals of `std`, so it was accepted in this case. That's just how this particular conversation resolved; I wouldn't necessarily use that as a reason to remove support for older versions in other scenarios.

Right, but do note also that I explicitly called out 4e5068c35cef5e1b3366aa05de60023b9fcd98e6 for review, which drops quite a few code paths unrelated to that.

> It's really just a low-priority extra that a rather small group of users might appreciate.

I'm sympathetic to that view, but with some caveats:

* I don't think there is any valid justification for using Windows versions older than 7. It's just inexcusable and is a symptom of an person/organization that refuses to upgrade full stop. We'd be irresponsible to enable that behavior.
* I can understand that some places still run Windows 7-8 despite Microsoft's support policy because rolling out upgrades is slower than it should be in some organizations. I'm not as dead-set against 7-8 support as I am against XP and older.
* I do think it should at least be demonstrated *somehow* that these Zig users on EoL Windows versions actually exist before we add code to support them.

(These are just my personal views of course.)

Reply:
> I do think it should at least be demonstrated somehow that these Zig users on EoL Windows versions actually exist before we add code to support them.

I don't agree, because people still using Windows 7 will simply not use Zig when they discover it doesn't work on their machine and use C instead. That's worse for everyone involved, and certainly not serving the users.

Here are some examples of Zig on Windows 7: #18690 #8951 https://github.com/ziglang/zig/issues/7654#issuecomment-753608306

> I can understand that some places still run Windows 7-8 despite Microsoft's support policy because rolling out upgrades is slower than it should be in some organizations (but it does at least happen). I'm not as dead-set against 7-8 support as I am against XP and older.

At least as of 2021, Windows XP and older still sees widespread use in lots of infrastructure, such as public displays or ATMs:

https://www.bleepingcomputer.com/news/microsoft/its-windows-xps-20th-birthday-and-way-too-many-still-use-it/
https://scribe.rip/hackernoon/do-atms-running-windows-xp-pose-a-security-risk-you-can-bank-on-it-1b7817902d61

Organizations generally don't care about security unless there is a problem. Fixing security holes and keeping systems up to date is seen as costly and unnecessary: if it works, why change it? That's a short-sighted view, of course, but it is very prevalent. As a pragmatic successor to C, Zig has to take that into account.